### ISSUE TITLE
Context isolation in the VM

### ISSUE THREAD

**Issue by bobbinth** [maticnetwork/miden#231]
To support a transaction model similar to the one described in #192, we'll need to have the ability to place restrictions on what programs can do in certain contexts. To list some restrictions:

* Account functions should be able to create new notes, but should not be able modify notes after they have been created.
* Note scripts should be able to invoke account's functions, but should not be able to modify account storage/vault directly.
* Neither account functions nor note scripts should be able to access the data built during transaction prologue and epilogue (i.e., aggregate vaults of inputs/outputs).

Separately, we also want to have a set of environment variables (e.g., block height, account ID etc.) to be accessible to account functions and note scripts. These variables should be read-only - i.e., neither account functions nor note scripts should be able to change their values.

### Potential construction

One potential way to support this functionality is by having isolated contexts for different types of programs. Let's call this contexts "spaces". To support the above, we could have 3 such spaces:

* **Kernel space** - this is where transaction prologue, epilogue etc. would be executed.
* **Account space** - this is where functions belonging to an account would be executed.
* **Note space** - this is where note script would be executed.

These spaces would have isolated memories (our memory model already supports this). We'll also need to find a way to place restrictions on functions which are callable from each space. For example:

* For read-only variables, there could be procedures like `get_account_id`, `get_block_height` etc. - these would be callable from any space.
* For access to account storage, there could be procedures `get_storage_item` and `set_storage_item` - calling these from the note space should not be allowed.
* For access to account vault, there could be procedures like `get_asset` and `set_asset`- calling these from the note space should not be allowed.
* For creating notes, there could be a procedure like `create_note` - calling it from the note space should not be allowed.

### Potential implementation

Here is a very rough sketch of how the above functionality can be implemented within the VM.

First, we could add two columns to the execution trace to track which context we are in - e.g. $c_0$ and $c_1$ (there could also be a way to do it with only one column, but for simplicity, I'll use two). When $c_0 = 1$, we are in the kernels space, when $c_1 = 1$ we are in the account space, if both are zeros, then we are in the note space.

Next, we could introduce a new opcode - something like `syscall` which would be used to call functions in the kernel space. This opcode could be used as follows:
```
syscall.[kernel function hash]
```
When the VM executes the above it would do the following:
1. Make sure that specified function hash actually belongs to the set of available kernel functions. This can be done by verifying a Merkle path against a root of kernel code. The root of kernel code would be static (i.e., hard-coded into the AIR).
2. Set memory context to `0` - which would be the context ID for kernel space.
3. Set $c_0$ to `1` and $c_1$ to `0`.
4. Push `1` onto the stack if we are currently in account space, and `0` otherwise. This way, the invoked function would know which space the call came from. (calling kernel functions from kernel space would not be allowed - but that's not a problem, within kernel space we could use regular procedures).
5. Then execute the actual code of the function.

![image](https://user-images.githubusercontent.com/43513081/171490432-90d1dc0b-4377-420c-91d0-1ec5bd79f25f.png)

All of the functions described previously could be kernel function. For example, `get_account_id` could be a kernel function which reads account ID from kernel memory and puts it onto the stack. For functions like `get_storage_item`, the the first thing would be to check if the function was invoked from the account space (and fail otherwise), and then do a lookup into the account's storage Merkle tree, and put the node value onto the stack.

To support calling account functions, we could add another opcode - something like `actcall`. This opcode could be used as follows:
```
actcall.[account function hash]
```
When the VM executes the above it would do the following:
1. Make sure that the specified function hash actually belongs to the account functions. This can be done by verifying a Merkle path against account's code root.
2. Set memory context to some new unique value (could be the same as code block ID, as these guaranteed to be unique). This would imply that every call to an account's functions would be executed in separate memory space.
3. Set $c_0$ to `0` and $c_1$ to `1`.
4. Then execute the actual code of the function.

![image](https://user-images.githubusercontent.com/43513081/171490126-3de6e4bf-ec29-421c-8c3b-9b82b1de6553.png)

There is one complication here: unlike with `syscall`, we don't have a static root for account code (i.e., every account could have a different code root). Thus, we can't hardcode account code root into the AIR. So, we would need to have a place from where the VM could retrieve account code root. The simplest solution is to dedicate 4 trace columns for it. This is, however, rather wasteful - values in these columns would be set only once for a given transaction and would remain the same for the duration of the entire execution trace. There might be a better way to handle this - but I haven't come up with anything yet.

### Summary
To summarize the above approach:
1. We need to add potentially 6 new columns to the VM: $c_0$, $c_1$, and 4 columns to hold account code root.
2. We need to introduce two new opcodes `syscall` and `actcall` and implement their semantics as described above.
3. We need to implement context switching (e.g., when executing `syscall` and `actcall` but also when returning from these functions).

Open questions:
1. Is there a better way to provide account code root to the VM?
2. Is there a better way in general to handle this problem. For example, suppose we add read-only memory to the VM, would this change things?

**Comment by bobbinth** [comment]
After thinking a little more about this, I think there is a better (and much simpler) way of supporting the functionality described above.

First, instead of having 3 spaces, we have 2:

1. Kernel space.
2. User space.

Every program starts execution in the Kernel space (memory context = 0). To go into the user space we use a `call` operation. This operation can be used as follows:
```
call.[function hash]
```
When the VM executes the above it would do the following:
1. Set memory context to some new unique value. This would mean that the function code would not have access to kernel memory.
2. Then execute the actual code of the function.

To make calls from user space into kernel space we would have a `syscall` operation. This operation can be used as follows:
```
syscall.[kernel function hash]
```
When the VM executes the above it would do the following:
1. Make sure that specified function hash actually belongs to the set of available kernel functions. This can be done by verifying a Merkle path against a root of kernel code. The root of kernel code would be static (i.e., hard-coded into the AIR).
2. Set memory context to 0 - which would be the context ID for kernel space.
3. Put the hash of the currently executing function onto the stack.
4. Then execute the actual code of the function.

Since the called kernel function will have access to the hash of the caller, it can check whether the caller belongs to the account code root (i.e., by verifying a Merkle path), and thus infer the privilege level of the caller.

A few open questions for this approach:
1. We need to keep the hash of the currently executing function somewhere so that it can be pushed onto the stack. The simple way is to have 4 dedicated columns for this. The `call` operation would set the values in these columns to the hash of the invoked function, and when the function returns, the values would be set to the caller's hash. There might be other ways of doing it which (i.e., via a "virtual table") - but I'm not sure if they are worth it.
2. If we have 4 dedicated columns for current function hash - do we need to push the function hash onto the stack when `syscall` is executed? Maybe executing `syscall` wouldn't change the contents of these columns, and then we add another operation - e.g., something like `load.env.croot` - which would load their contents onto the stack?
3. Handling of the `fmp` register would require more thinking. At the minimum, we'll need to keep track of its value in every context - but i'm not sure if anything more than that is needed. Maybe it is worth revisiting the idea of storing `fmp` value in memory, rather than in a dedicated column.

**Comment by bobbinth** [comment]
Another drawback of the above approach is hardcoding of kernel root. For one, it makes the design less flexible. Also, as discussed in https://github.com/maticnetwork/miden/discussions/248#discussioncomment-2959921 we'd actually need two different kernels.

Supporting non-hard-coded kernels can be done in a couple of ways.

One way is to have 4 columns which contain the root of the kernel code. values in these columns would never change during program execution, and we can enforce boundary constraints against the initial values to specify which kernel to use. The obvious drawback of this approach is that we need to use 4 extra columns. Another drawback is that we need to use Merkle path verification procedure for every kernel call (and these have non-negligible costs).

Another way is to have a sort of read-only memory (ROM) which would contain all kernel functions. The state of this ROM could be enforced via boundary constraints. We probably won't save a lot in terms of number of extra columns (we'd probably need 3 - 4 extra columns for this approach as well), but checking if some value is a root of a kernel function should become much cheaper. Also, if we do have (ROM) it could be use for other purposes (e.g., maybe for recording powers of 2 up to 2^63, and in that case, we might be able to get rid of the power of two processor).

One other thing to figure out with this latter approach is how to initialize this ROM in such a way that would not reveal which kernel functions were called and how many times. I think this can be done with regular PLOOKUP approach - but we'll need to nail down the details.

**Comment by bobbinth** [comment]
To summarize my current thinking:

1. We should probably add 4 columns which would contain hash of the currently executing function. These columns would behave as follows:
  a. Initial values of these columns would be set to ZEROs.
  b. The values in these columns would be set when `call` operation is executed, and would be reset to previous values when we return from a function call.
  c. When `syscall` operation is executed, the values in these columns do not change.
2. We should add 3 new operations to the VM:
  a. `call`
  b. `syscall` - inside a `syscall` it should not be possible to execute a `call` or another `syscall`. We'll probably need to add another column to help enforce this.
  c. An operation which loads the values from the above 4 columns onto the stack.
3. We should probably add ROM component which can be initialized with functions of a specific kernel. This will probably require another 3 - 4 columns.

Open questions:
* How to implement this ROM? And can we implement it in such a way as to also use it instead of power of two processor and maybe other things?
* How to handle `fmp` register?
* And more generally, how would this impact decoder constraints?

**Comment by bobbinth** [comment]
It maybe be possible to implement kernel ROM without adding any extra columns. Here is how I think this can be done.

First, let's start with the most basic functionality. Let's denote $p_0$ as the running product column which keeps track of all kernel function calls. When we make a kernel call the value of $p_0$ is divided by the value representing the root of the called function (4 field elements). Specifically:

$$
p_0' \cdot (\alpha_0 + \sum_{i=0}^3 (\alpha_{i + 1} \cdot r_i)) = p_0
$$

Where, $r_i$ are the registers containing elements of kernel function root.

Then, to ensure that `syscall` operation executed only valid kernel function, the prover would initialize $p_0$ with an aggregate value of all called functions. This way, by the time the program finishes executing, $p_0$ would must be equal to $1$.

The above approach works and it is actually very efficient: we don't need any additional columns besides $p_0$ (columns $r_0$, ... $r_3$ are already a part of the decoder). But it does have a very undesirable property: the verifier needs to know the initial value of $p_0$, and this reveals which kernel functions have been called by the program. This, in turn, compromises privacy.

We can address this by having a dedicated ROM table which will consist of 6 columns as shown below:

![image](https://user-images.githubusercontent.com/43513081/178194489-3156b5ad-db8f-45bc-812c-bd91a9443ac4.png)

The meaning of columns in the above is as follows:
* Column $s_0$ specifies whether the value in the row should be included into column $p_0$.
* $addr$ is the row address column which starts out at $0$ and must either remain the same or be incremented by $1$ with every row.
* $q_0, ..., q_3$ are the columns containing roots of kernel functions. In the example above, the kernel consists of 3 functions $a$, $b$, and $c$. Values in these column can change only when the value in the $addr$ column changes. If $add$ column remains the same, values in $q$ columns must also remain the same.

We also need to add another running product column $p_1$. This column will keep track of all *unique* kernel function roots. The values in this column will be updated only when the value in the address column change.

To describe constraints against $p_0$ and $p_1$ more formally:

Let's define $\Delta addr = addr' - addr$. Let's also compute row value included into $p_0$ as:

$$
v_0 = \alpha_0 + \sum_{i=0}^3 (\alpha_{i + 1} \cdot q_i)
$$

And row value included into $p_1$ as:

$$
v_1 = \alpha_0 + \alpha_1 \cdot addr + \sum_{i=0}^3 (\alpha_{i + 2} \cdot q_i)
$$

Then the constraint against $p_0$ would be (this is actually only one side of the constraint; the other side, the one which divides values out of $p_0$ is omitted):

$$
p_0' = p_0 \cdot (s_0 \cdot v_0 - 1 + s_0)
$$

Thus, when $s_0 = 0$, the above reduces to $p_0' = p_0$, but when $s_0=1$, the above becomes $p_0' = p_0 \cdot v_0$.

The constraint against $p_1$ would be:

$$
p_1' = p_1 \cdot (\Delta addr  \cdot v_1 - 1 + \Delta addr) 
$$

Thus, when $\Delta addr = 0$, the above reduces to $p_1'=p1 \text{, but when } \Delta addr = 1$, the above becomes $p_1' = p_1 \cdot v_1$.

We also need to impose boundary constraints to make sure that:
* $p_0$ is equal to $1$ at the first and last steps of the execution trace.
* $p_1$ is qual to $1$ at the first row and to the product of all unique kernel functions in the last row.

The last boundary constraint means that the verifier only needs to know which kernel was used, but doesn't need to know which functions were invoked within the kernel.

In terms of costs, the above structure requires 6 column in the main execution trace and 2 columns in the auxiliary trace segment. However, in the context of the VM, we don't need to allocate any new columns and can instead reuse columns we already have. Specifically:

* The 6 columns for the table could go into the auxiliary co-processor table. We would need to introduce a new 4-column prefix, but otherwise, things remain the same.
* For $p_0$, we can reuse the auxiliary table bus column.
* For $p_0$, we can reuse the column used to trace the sibling table for Merkle root computations.

If we go with this approach, the new co-processor trace would look something like this:

![image](https://user-images.githubusercontent.com/43513081/178200150-b9af8737-0857-4160-978b-5ac4a9f6983c.png)

**Comment by grjte** [comment]
Just a thought (and I haven't thought about other implications) - If we wanted to, couldn't we prevent the verifier from knowing even which kernel was used by having a ROM table for each kernel. If the kernel doesn't have too many functions then it wouldn't add that many rows in the case where is not used

**Comment by bobbinth** [comment]
> I assume this was meant to be `p_1` which would share the sibling table aux column (the `p_1` column in the hasher)

Yes - correct. Fixed!

> To clarify - each unique function in the given kernel gets at least one row in the table, right? So the ROM table contains all functions in the kernel, but some of them possibly multiple times.

Yes, correct. The number of times is directly related to how many times the function is called from the executed program.

> How will we enforce that they prover doesn't choose functions it's not allowed to use?

* The final value of $p_1$ will have a boundary constraint which fixes the kernel used in a given computation.
* By design, $p_0$ column can contain function roots which come from the specified kernel, and nothing else.
* Whenever a kernel function is called in the VM (via `syscall`), we will divide $p_0$ by the value representing the root of the called function (this will be enforced by other constraints). Thus, if the prover uses some function not in the kernel, value of $p_0$ will not equal $1$ at the end of the trace (and this will be enforced by a boundary constraint).

There is one potential issue with the last point above: if at the time of `syscall` we also need to initiate hash computation, this will also require modification of $p_0$ column. This will likely push the degree of constraints agains $p_0$ beyond 9. So, either we'll need to design `syscall` in such a way that it doesn't initiate hash computations, or we'd need to add an extra auxiliary trace column.

> Just a thought (and I haven't thought about implications yet) - If we wanted to, couldn't we prevent the verifier from knowing even which kernel was used by having ROM tables for multiple kernels? 

I think we probably want the verifier to know which kernel was used (at least in some cases). For example, if there is a separate kernel for transaction processing and transaction aggregation, we probably don't want to make functions from transaction aggregation kernel available during transaction processing and vice-versa.

**Comment by maxgillett** [comment]
On a similar note, I believe we need a mechanism to store data that is shared between kernel calls. For example, in the transaction kernel, vault contents of the prologue needs to be stored and compared with those of the epilogue, and in the block kernel, note consumption need to be tracked as transactions are verified to avoid double spending. One solution is to introduce a Kernel Memory coprocessor, similar to the read-write memory, but without a context column and only accessible to kernel calls.  I can write in more detail what changes need to be made to the constraints to achieve this if there is not a better way of going about this.

**Comment by bobbinth** [comment]
> One solution is to introduce a Kernel Memory coprocessor, similar to the read-write memory, but without a context column and only accessible to kernel calls.

I was actually thinking we could just set `context = 0` for kernel memory. This way, no additional structure is needed - we'll just use the same memory chiplet (new name for "coprocessor") as we have now.

**Comment by bobbinth** [comment]
Closed by #375, #377, #403, #412, #443, #451, and #475.
